shepherd/Tester.py

```diff
@@ -0,0 +1,691 @@
+    file along with everything imported above is available to this function.
+    Keep this in mind, since this function is called on potentially unsafe user
+    entered code.
+    """
+    return globals()[name]
+
+
+def get_attr_from_name(source, name):
+    """
+    A helper function used to get an attribute of a certain name from a given
+    class (source).
+    """
+    if not isinstance(source, type):
+        raise Exception('{} is not a class.'.format(source))
+    return getattr(source, name)
+
+
+def parse_header(header):
+    """
+    A helper function that translates the headers in format
+    <LCM_targer>.<header> to the string representation stored in Utils.py.
+    Enforces the syntax for referencing the header, as well as the existance of
+    the header in Utils.py.
+    """
+    parts = header.split('.')
+    if len(parts) != 2:
+        raise Exception('{} is invalid.'.format(header))
+    ex = None
+    try:
+        klass = get_class_from_name(parts[0])
+    except KeyError:
+        ex = Exception(
+            '{} is not recognized. Make sure this is a class of headers in Utils.py'.format(parts[0]))
+    finally:
+        if ex:
+            raise ex
+    ex = None
+    try:
+        name = get_attr_from_name(klass, parts[1])
+    except AttributeError:
+        ex = Exception('{} is not recognized. Make sure this is a header in {} in Utils.py'.format(
+            parts[1], parts[0]))
+    finally:
+        if ex:
+            raise ex
+    return name
+
+
+def execute_python(script):
+    """
+    A helper function that executes a python expression in the context of the
+    local scipt enviroment.
+    """
+    global LOCALVARS
+    exec(script, LOCALVARS)
+
+
+def evaluate_python(token):
+    """
+    A helper function that evaluates a token against the local scipt enviroment.
+    """
+    global LOCALVARS
+    return eval(token, LOCALVARS)
+
+
+def tokenize_wait_exp(expression):
+    """
+    The function that parses WAIT statements.
+    Enforces syntax and returns a dictionary of the deconstructed WAIT statement.
+    """
+    global TARGET
+    original_expression = expression
+
+    def helper_min(a, b):
+        """
+        A helper function that returns the min between a and b, but considers -1
+        to be an unacceptable return value.
+        """
+        if a == -1 and b == -1:
+            return None
+        elif a == -1:
+            return b
+        elif b == -1:
+            return a
+        else:
+            return min(a, b)
+    tokens = expression.split('FROM')
+    if len(tokens) != 2:
+        raise Exception('expected to find FROM after <header> and before <target> in WAIT statement: WAIT {}'.format(
+            original_expression))
+    header = parse_header(remove_outer_spaces(tokens[0]))
+    expression = tokens[1]
+    if remove_outer_spaces(expression[0:helper_min(expression.find(' SET '), expression.find(' WITH '))].split('.')[0]) != 'LCM_TARGETS':
```

Comment:
ok, I'll change this, but the actual correct way to do this would be to use token objects with metadata and a regex tokenizer to break down the line before it gets here...

Reply:
done